libs/fluid-elements/virtual-scroll-container/src/lib/virtual-scroll-container.ts

```diff
@@ -0,0 +1,582 @@
+            this._cachedItems.get(this.trackByFn(this._items[itemIndex]))!
+              .height *
+              VIRTUAL_SCROLL_THRESHOLD >=
+            entry.rootBounds!.top
+        ) {
+          // Observed item is at the start of the list, the item is already below the
+          // bottom edge of the container, hence the observer would not trigger again
+          // when scrolling further up
+          this._clearVirtualScrollBufferObserver();
+          this._scrollState.updateStartScrollBuffer();
+          this._renderVirtualScrollItems();
+          this._updatePlaceholders();
+        }
+
+        this._observedItemsIndices.add(itemIndex);
+      } else {
+        // Request buffer update and render updated items
+        if (entry.isIntersecting) {
+          this._clearVirtualScrollBufferObserver();
+
+          if (itemIndex === this._scrollState.bufferStartObservedItemIndex) {
+            this._scrollState.updateStartScrollBuffer();
+          } else {
+            this._scrollState.updateEndScrollBuffer();
+          }
+
+          this._renderVirtualScrollItems();
+          this._updatePlaceholders();
+        }
+      }
+    });
+  }
+
+  /** Clears items observed for handling the virtual scroll buffer */
+  private _clearVirtualScrollBufferObserver(): void {
+    this._virtualScrollBufferObserver.disconnect();
+    this._observedItemsIndices.clear();
+  }
+
+  /** Checks, if the rendered items are filling up the available space */
+  private _isInitialRenderDone(): boolean {
+    if (
+      this._scrollContainer.clientHeight <
+        this._scrollItemsContainer.clientHeight ||
+      this._scrollState.renderedItemsRange.last === this.items.length - 1
+    ) {
+      // The container is completely filled up or the last item is rendered
+      this._initialRenderDone = true;
+    }
+
+    return this._initialRenderDone;
+  }
+
+  /**
+   * Calculates average item height of the currently rendered items
+   * and sets the height of the placeholders
+   */
+  private _updatePlaceholders(): void {
+    // Store the current scrollTop and reset it after setting the height to avoid jumping
+    const currentScroll = this._scrollContainer.scrollTop;
+
+    // Recalculate the size of the placeholders
+    this._scrollState.calculatePlaceholderSize();
+    // Set height of the placeholders above and below the rendered items
+    // to correctly display the scroll bar
+    this._setPlaceholders();
+
+    this._scrollContainer.scrollTop = currentScroll;
+  }
+
+  /** Sets the height of the placeholder */
+  private _setPlaceholders(): void {
+    if (this._items.length) {
+      this._scrollStartPlaceholder.style.height = `${this._scrollState.startPlaceholderSize}px`;
+      this._scrollEndPlaceholder.style.height = `${this._scrollState.endPlaceholderSize}px`;
+    } else {
+      this._scrollStartPlaceholder.style.height = `0`;
+      this._scrollEndPlaceholder.style.height = `0`;
+    }
```

Comment:
This would be an example for an early exit
```suggestion
    if (this._items.length) {
      this._scrollStartPlaceholder.style.height = `${this._scrollState.startPlaceholderSize}px`;
      this._scrollEndPlaceholder.style.height = `${this._scrollState.endPlaceholderSize}px`;
      return;
    } 
    
    this._scrollStartPlaceholder.style.height = `0`;
    this._scrollEndPlaceholder.style.height = `0`;
```